File: pixie/rendering.py
import hashlib
from io import StringIO
import json
import os
from pathlib import Path
import re
from tabnanny import check
import requests
import tempfile
import logging
import boto3
from typing import List
from git import Repo
from giturlparse import parse

from jinja2 import Environment, ChainableUndefined
from jinja2 import FileSystemLoader
from jinja2.nativetypes import NativeEnvironment
from ruamel.yaml import YAML
from termcolor import colored

from pixie.context import PixieContext


_log = logging.getLogger(name=__name__)


class AwsUtils(object):
    def get_parameter(self, name, profile_name=None):
        try:
            session = boto3.Session(
                profile_name=profile_name
            )

            ssm_client = session.client('ssm')
            param = ssm_client.get_parameter(Name=name)
            return param['Parameter']['Value']
        except Exception as ex:
            _log.debug(ex)
        return ''


class GitUtils(object):
    def remote_info(self, path, remote_name):
        try:
            repo = Repo(path)
            remote = repo.remote(remote_name)
            return parse(remote.url, check_domain=False)
        except Exception as ex:
            _log.debug(ex)
        return {}
    
    def owner_repo(self, path, remote_name):
        try:
            repo = Repo(path)
            remote = repo.remote(remote_name)
            u = parse(remote.url, check_domain=False)
            return u.owner + '/' + u.repo
        except Exception as ex:
            _log.debug(ex)
        return ''


class RenderUtils(object):  # pylint: disable=R0903
    """Template utilities."""

    def __init__(self, context) -> None:
        super().__init__()

        self.context = context
        self.colored = colored

    @classmethod
    def read_file(cls, path, parse=False):
        """Used to read a file and return its contents."""

        with open(path, 'r') as file_handle:
            if parse:
                parser = get_parser(path)
                return parser.load(file_handle)
            else:
                return file_handle.read()

    @classmethod
    def read_json(cls, path):
        """Used to read a JSON file and return its contents."""

        with open(path, 'r') as file_handle:
            return json.load(file_handle)

    @classmethod
    def read_yaml(cls, path):
        """Used to read a YAML file and return its contents."""
        yaml = YAML()
        with open(path, 'r') as file_handle:
            return yaml.load(file_handle)
    
    @classmethod
    def random_string(cls, length=16, special_chars=''):
        chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ23456789" + special_chars
        from os import urandom

        return "".join(chars[c % len(chars)] for c in urandom(length))

    def checksum(cls, file):
        with open(file, 'rb') as fhd:
            return hashlib.md5(fhd.read()).hexdigest()

    def download_file(cls, url, checksum):
        file = tempfile.mktemp()
        resp = requests.get(url)
        if resp.status_code == 200:
            with open(file, 'w') as fhd:
                fhd.write(resp.text)
            actual_checksum = cls.checksum(file)
            if checksum != actual_checksum:
                raise RuntimeError('checksum mismatch for %s (actual=%s, expected=%s)',
                    file, actual_checksum, checksum)
        else:
            raise RuntimeError('failed to download file (code=%s)' % resp.status_code)
        return file

    def path_exists(cls, path):
        return os.path.exists(os.path.expanduser(path))
    
    def join_arrays(cls, *arrays):
        output_array = []
        for array in arrays:
            output_array.extend(array)
        return output_array
    
    def dirs(cls, path):
        return [x.name for x in Path(path).iterdir() if x.is_dir]

    def files(cls, path):
        return [x.name for x in Path(path).iterdir() if x.is_file]
    
    def glob(cls, path):
        return Path(path).glob(path)


def format_list(value, format='{value}'):
    for idx, x in enumerate(value):
        value[idx] = format.format(value=value[idx], index=idx)
    return value


def yaml_format(value, **kw):
    if value is None:
        return 'null'
    yaml = YAML()
    stream = StringIO()
    yaml.dump(value, stream, **kw)
    return stream.getvalue().rstrip()


def json_format(value):
    if value is None:
        return 'null'
    return json.dumps(value)


def join_path(value, added_path):
    if value is None:
        return 'null'
    return os.path.join(value, added_path)


def get_parser(path):
    ext = os.path.splitext(path)[1]
    if ext == '.yaml' or ext == '.yml':
        yaml = YAML()
        return yaml
    elif ext == '.json':
        return json
    else:
        exit('Parser format not supported: %s' % ext)


def get_utils(context):
    return dict(
        utils=RenderUtils(context),
        aws=AwsUtils(),
        git=GitUtils(),
        path=os.path,
        tempfile=tempfile
    )


def render(template_name, context, template_dir):
    """Used to render a Jinja template."""

    env = Environment(loader=FileSystemLoader(template_dir), variable_start_string='${{', variable_end_string='}}', keep_trailing_newline=True, undefined=ChainableUndefined)
    add_filters(env)
    utils = get_utils(context)

    template = env.get_template(template_name)

    return template.render(context=context, **utils, **context)


def add_filters(env):
    env.filters['formatlist'] = format_list
    env.filters['yaml'] = yaml_format
    env.filters['json'] = json_format
    env.filters['join_path'] = join_path


def render_value(text, context: PixieContext):
    """Used to render a Jinja template."""

    if text is None:
        return None

    if not isinstance(text, str):
        return text

    if '\n' in text:
        return render_text(text, context)

    env = NativeEnvironment(variable_start_string='${{', variable_end_string='}}', undefined=ChainableUndefined)
    add_filters(env)

    utils = get_utils(context)

    template = env.from_string(text)

    return template.render(context=context, **utils, **context)


def render_text(text, context: PixieContext):
    """Used to render a Jinja template."""

    if text is None:
        return None

    env = Environment(variable_start_string='${{', variable_end_string='}}', keep_trailing_newline=True, undefined=ChainableUndefined)
    add_filters(env)
    utils = get_utils(context)

    template = env.from_string(text)

    return template.render(context=context, **utils, **context)


def _render_value(value, context: PixieContext, exclude_keys: List[str]):
    if isinstance(value, str):
        return render_value(value, context)
    elif isinstance(value, list):
        v_list: list[str] = []
        for x in value:
            v_list.append(_render_value(x, context, []))
        return v_list
    elif isinstance(value, dict):
        opts = value.copy()
        for k, v in opts.items():
            if k in exclude_keys:
                opts[k] = v
            else:
                opts[k] = _render_value(v, context, [])
        return opts
    else:
        return value

def render_options(options: dict, context: PixieContext, exclude_keys=[]):
    return _render_value(options, context, exclude_keys)

def render_token_file(path: str, tokens: dict):
    """Used to render a Token File."""

    with open(path, 'r') as file_handle:
        content = file_handle.read()

    return render_tokens(content, tokens)

def render_tokens(content: str, tokens: dict):
    """Used to render a Token File."""

    for token in tokens:
        content = content.replace(token, tokens[token])
    return content

File: src/rendering.ts
import * as nunjucks from 'nunjucks';
import * as fs from 'fs';
import * as path from 'path';
import * as crypto from 'crypto';
import * as os from 'os';
import { PixieContext } from './context';
import * as yaml from 'js-yaml';

export class RenderUtils {
  private context: PixieContext;

  constructor(context: PixieContext) {
    this.context = context;
  }

  static readFile(filePath: string, parse: boolean = false): any {
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      if (parse) {
        const ext = path.extname(filePath).toLowerCase();
        if (ext === '.json') {
          return JSON.parse(content);
        } else if (ext === '.yaml' || ext === '.yml') {
          return yaml.load(content);
        }
      }
      return content;
    } catch (error) {
      console.error(`Error reading file ${filePath}:`, error);
      return null;
    }
  }

  static readJson(filePath: string): any {
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      return JSON.parse(content);
    } catch (error) {
      console.error(`Error reading JSON file ${filePath}:`, error);
      return null;
    }
  }

  static readYaml(filePath: string): any {
    try {
      const content = fs.readFileSync(filePath, 'utf8');
      return yaml.load(content);
    } catch (error) {
      console.error(`Error reading YAML file ${filePath}:`, error);
      return null;
    }
  }

  static randomString(length: number = 16, specialChars: string = ''): string {
    const chars = "abcdefghijklmnopqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ23456789" + specialChars;
    const randomBytes = crypto.randomBytes(length);
    return Array.from(randomBytes, byte => chars[byte % chars.length]).join('');
  }

  checksum(filePath: string): string {
    try {
      const content = fs.readFileSync(filePath);
      return crypto.createHash('md5').update(content).digest('hex');
    } catch (error) {
      console.error(`Error calculating checksum for ${filePath}:`, error);
      return '';
    }
  }

  pathExists(filePath: string): boolean {
    return fs.existsSync(filePath);
  }

  joinArrays(...arrays: any[][]): any[] {
    return arrays.flat();
  }

  dirs(dirPath: string): string[] {
    try {
      return fs.readdirSync(dirPath).filter(item => {
        return fs.statSync(path.join(dirPath, item)).isDirectory();
      });
    } catch (error) {
      console.error(`Error reading directory ${dirPath}:`, error);
      return [];
    }
  }

  files(dirPath: string): string[] {
    try {
      return fs.readdirSync(dirPath).filter(item => {
        return fs.statSync(path.join(dirPath, item)).isFile();
      });
    } catch (error) {
      console.error(`Error reading directory ${dirPath}:`, error);
      return [];
    }
  }
}

export class GitUtils {
  remoteInfo(repoPath: string, remoteName: string): any {
    // Simplified git remote info - in a real implementation you'd use a git library
    console.debug(`Getting remote info for ${repoPath}/${remoteName}`);
    return {};
  }
}

export function formatList(value: any[], format: string = '{value}'): string {
  if (!Array.isArray(value)) {
    return String(value);
  }
  return value.map(item => format.replace('{value}', String(item))).join(', ');
}

export function yamlFormat(value: any): string {
  return yaml.dump(value);
}

export function jsonFormat(value: any): string {
  return JSON.stringify(value, null, 2);
}

export function joinPath(basePath: string, addedPath: string): string {
  return path.join(basePath, addedPath);
}

function getUtils(context: PixieContext) {
  return {
    utils: new RenderUtils(context),
    git: new GitUtils(),
    path: path,
    os: os
  };
}

function createEnvironment(variableStart = '${{', variableEnd = '}}', trimBlocks = true): nunjucks.Environment {
  const env = new nunjucks.Environment(null, {
    trimBlocks,
    lstripBlocks: true,
    tags: {
      variableStart,
      variableEnd
    }
  });

  // Add custom filters
  env.addFilter('formatlist', formatList);
  env.addFilter('yaml', yamlFormat);
  env.addFilter('json', jsonFormat);
  env.addFilter('join_path', joinPath);

  return env;
}

export function renderValue(text: any, context: PixieContext): any {
  if (text === null || text === undefined) {
    return text;
  }

  if (typeof text !== 'string') {
    return text;
  }

  if (text.includes('\n')) {
    return renderText(text, context);
  }

  const env = createEnvironment();
  const utils = getUtils(context);
  
  // Convert context Map to object for template rendering
  const contextObj: Record<string, any> = {};
  for (const [key, value] of context.entries()) {
    contextObj[key] = value;
  }

  try {
    const template = nunjucks.compile(text, env);
    return template.render({ context: contextObj, ...utils, ...contextObj });
  } catch (error) {
    console.error('Error rendering value:', error);
    return text;
  }
}

export function renderText(text: string | null, context: PixieContext): string | null {
  if (text === null || text === undefined) {
    return text;
  }

  const env = createEnvironment();
  const utils = getUtils(context);
  
  // Convert context Map to object for template rendering
  const contextObj: Record<string, any> = {};
  for (const [key, value] of context.entries()) {
    contextObj[key] = value;
  }

  try {
    const template = nunjucks.compile(text, env);
    return template.render({ context: contextObj, ...utils, ...contextObj });
  } catch (error) {
    console.error('Error rendering text:', error);
    return text;
  }
}

function _renderValue(value: any, context: PixieContext, excludeKeys: string[] = []): any {
  if (typeof value === 'string') {
    return renderValue(value, context);
  } else if (Array.isArray(value)) {
    return value.map(item => _renderValue(item, context, []));
  } else if (typeof value === 'object' && value !== null) {
    const result: Record<string, any> = {};
    for (const [key, val] of Object.entries(value)) {
      if (excludeKeys.includes(key)) {
        result[key] = val;
      } else {
        result[key] = _renderValue(val, context, []);
      }
    }
    return result;
  }
  return value;
}

export function renderOptions(options: Record<string, any>, context: PixieContext, excludeKeys: string[] = []): Record<string, any> {
  return _renderValue(options, context, excludeKeys);
}

export function renderTokenFile(filePath: string, tokens: Record<string, string>): string {
  const content = fs.readFileSync(filePath, 'utf8');
  return renderTokens(content, tokens);
}

export function renderTokens(content: string, tokens: Record<string, string>): string {
  let result = content;
  for (const [token, replacement] of Object.entries(tokens)) {
    result = result.replace(new RegExp(token, 'g'), replacement);
  }
  return result;
}